### PR TITLE
Fix the assignment of multiple quantification window coordinates

### PR DESCRIPTION
**Pull request by Colelyman**
This PR fixes the assignment of multiple quantification window coordinates (qwc) when provided with multiple amplicons.

Before this PR, if you provided multiple amplicons and specified multiple qwc's the qwc for the second amplicon would be in reference to the first amplicon. With this PR, the qwc for the second (or third, fourth, etc.) amplicon is not with respect to the first amplicon.

One remaining question @kclem before we merge this in, if multiple amplicons are provided, but only one qwc is specified then the qwc for the second amplicon is adjusted to be in reference to the first amplicon (via s1inds). Is this desired behavior?

### CODE REVIEW

**Comment by mbowcut2** [comment]
@Colelyman merged master in and this passes all integration tests. I think we're ready for review and merging. LMK if you want to go over test cases together, want more coverage, etc.

File: tests/unit_tests/test_CRISPRessoCORE.py [hunk 1]
```diff
@@ -93,6 +93,105 @@ def test_get_consensus_alignment_from_pairs():
...
+    # 5 bp deletion of entire qw
+    quant_window_coordinates = '1-4_7-10'
+    s1inds = [0, 1, 2, 3, 4, 5, 6, 6, 6, 6, 6]
+    assert CRISPRessoCORE.get_cloned_include_idxs_from_quant_window_coordinates(quant_window_coordinates, s1inds) == [1, 2, 3, 4]
+
+# NOTE: include_idxs should be exactly the numbers as specificed in the string, inclusively
+# For example: 20-30_175-185 --> [20, 21, ..., 30, 175, 176, ..., 185]
+    
+# BUG: you can't include the last base in the QW or you get an out of range error -- this is extant beyond the function tested here.
+
```
Comment (Colelyman) [inline tests/unit_tests/test_CRISPRessoCORE.py:194]: Could this be addressed in the PR? Or would it be better addressed in a separate PR?

File: tests/unit_tests/test_CRISPRessoShared.py [hunk 2]
```diff
@@ -21,3 +21,34 @@ def test_get_mismatches():
...
+    to_sequence = 'TTC-T'
+    from_sequence = 'TTCGT'
+    s1inds_gap_left, s1inds_gap_right = CRISPRessoShared.get_relative_coordinates(to_sequence, from_sequence)
+    assert s1inds_gap_left == [0, 1, 2, 2, 3] # should this be [0, 1, 2, 2, 4]?
+    assert s1inds_gap_right == [0, 1, 2, 3, 3] # should this be [0, 1, 2, 4, 4]?
+    assert from_sequence[0] == to_sequence[s1inds_gap_left[0]]
+    assert from_sequence[1] == to_sequence[s1inds_gap_left[1]]
+    assert from_sequence[2] == to_sequence[s1inds_gap_left[2]]
+    assert from_sequence[2] == to_sequence[s1inds_gap_left[2]]
+    # assert from_sequence[4] == to_sequence[s1inds_gap_left[4]] # this should accept, right?
```
Comment (Colelyman) [inline tests/unit_tests/test_CRISPRessoShared.py:42]: @mbowcut2 do you have an opinion of the questions in the comments here?

File: CRISPResso2/CRISPRessoCORE.py [hunk 3]
```diff
@@ -136,6 +136,80 @@ def get_n_reads_bam(bam_filename,bam_chr_loc=""):
...
+def get_cloned_include_idxs_from_quant_window_coordinates(quant_window_coordinates, idxs):
+    """Get the include_idxs from the quantification window coordinates, but adjusted according to s1ind.
+
+    Parameters
+    ----------
+    quant_window_coordinates: str
+        The quantification window coordinates, in the form "5-10_100-101", where
+        the "_" delimits separate ranges and the "-" delimits the range itself.
+    s1inds: list of int
+        The index values mapped to the amplicon from which this is being cloned.
```
Comment (Colelyman) [inline CRISPResso2/CRISPRessoCORE.py:193]: Could we update the name to match the parameter?

File: CRISPResso2/CRISPRessoCORE.py [hunk 4]
```diff
@@ -136,6 +136,80 @@ def get_n_reads_bam(bam_filename,bam_chr_loc=""):
...
+    list of int
+        The include_idxs to be used for quantification, which is the quantification
+        window coordinates expanded to include all individual indexes contained therein,
+        but adjusted according to s1inds.
+    """
+    include_idxs = []
+    for i, x in enumerate(idxs):
+        if i > 0:
+            if abs(idxs[i-1]) == x:
+                idxs[i] = -1 * abs(x)
```
Comment (Colelyman) [inline CRISPResso2/CRISPRessoCORE.py:206]: Here is the code where `s1inds` is generated (CRISPRessoShared.py:1967-1975)

```
s1inds = []
                s1ix = -1
                s2ix = -1
                for ix in range(len(fws1)):
                    if fws1[ix] != "-":
                        s1ix += 1
                    if fws2[ix] != "-":
                        s2ix += 1
                        s1inds.append(s1ix)
```

And instead of the `if i > 0`, could we just iterate over `idxs[1:]`? Maybe I'm not understanding entirely what this is doing...

File: tests/unit_tests/test_CRISPRessoCORE.py [hunk 5]
```diff
@@ -93,6 +93,105 @@ def test_get_consensus_alignment_from_pairs():
     assert ref_seq ==      "ATCGATCGAT"
     assert score == 50 #double check this score... should be 5/10
 
+
```
Comment (Colelyman) [inline tests/unit_tests/test_CRISPRessoCORE.py:96]: I LOVE the unit tests, thank you!

File: tests/unit_tests/test_CRISPRessoCORE.py (hunk 1 above)
Comment (Colelyman) [reply tests/unit_tests/test_CRISPRessoCORE.py:194]: PS, what would Uncle Bob think of this comment? ;)

**Review by Colelyman — COMMENTED** [review]
@mbowcut2 I think this is definitely on the right track and close to being merged in, let me know if you have any questions or objections to the comments!

File: tests/unit_tests/test_CRISPRessoCORE.py (hunk 1 above)
Comment (mbowcut2) [reply tests/unit_tests/test_CRISPRessoCORE.py:194]: I left this as a reminder. It worked! haha

So, this touches on a key issue: QWC's are (supposedly) 1-indexed. All of these changes assume that downstream calculations correctly utilize 1-index-based coordinates as input by the user, i.e. the user inputs `'1-5' --> include_idxs == [1, 2, 3, 4, 5]`. The test cases verify that these indexes are selected correctly.

So, I think the only necessary change to get the last char in the sequence would be to append one more index onto s1inds before we slice it.

File: CRISPResso2/CRISPRessoCORE.py (hunk 3 above)
Comment (mbowcut2) [reply CRISPResso2/CRISPRessoCORE.py:193]: sorry, which name and parameter?

Comment (Colelyman) [reply CRISPResso2/CRISPRessoCORE.py:193]: That's my bad, the parameter is idxs and the documentation says it is s1inds. The comment was to make them consistent.

File: tests/unit_tests/test_CRISPRessoCORE.py (hunk 1 above)
Comment (kclem) [reply tests/unit_tests/test_CRISPRessoCORE.py:194]: I think technically they [should be 0-indexed](https://github.com/pinellolab/CRISPResso2?tab=readme-ov-file#quantification-window-parameters) (but a value of 0 will disable the filter).

Comment (mbowcut2) [reply tests/unit_tests/test_CRISPRessoCORE.py:194]: Per the documentation "Indexes are 0-based, meaning that the first nucleotide is position 0." Thanks @kclem. So, there is no last bp bug.

However, including 0 in the qwc throws a `BadParameterException`, I'm assuming because 0 disables the filter. Should we keep this behavior as is?

Comment (Colelyman) [reply tests/unit_tests/test_CRISPRessoCORE.py:194]: @mbowcut2 I think it should behave such that if you have `-qwc 0-10_15-22,0` the first amplicon will have quantification windows `0-10` and `15-22` and then the second amplicon will have the quantification windo coordinates disabled (and therefore be set by where the sgRNA aligns).

Also, `-qwc 0` will outright disable it.

File: tests/unit_tests/test_CRISPRessoShared.py (hunk 2 above)
Comment (kclem) [reply tests/unit_tests/test_CRISPRessoShared.py:42]: The rule here is that from_sequence[i] matches to to_sequence[inds[i]]
Yes, should be [0,1,2,2,4] and [0,1,2,4,4]
Yes, `from_sequence[4] == to_sequence[s1inds_gap_left[4]]`

File: CRISPResso2/CRISPRessoCORE.py (hunk 3 above)
Comment (mbowcut2) [reply CRISPResso2/CRISPRessoCORE.py:193]: done

File: CRISPResso2/CRISPRessoCORE.py (hunk 4 above)
Comment (mbowcut2) [reply CRISPResso2/CRISPRessoCORE.py]: The goal here is to map: `[0,1,2,2,2,2,3] -> [0,1,2,-2,-2,-2,3]`. I just pushed a change that gets rid of `if i > 0`.

File: tests/unit_tests/test_CRISPRessoShared.py (hunk 2 above)
Comment (mbowcut2) [reply tests/unit_tests/test_CRISPRessoShared.py]: I don't understand how that could hold. Like in this test case:

`to_sequence = 'TTC-T'`
`from_sequence = 'TTCGT'`

`s1inds_gap_left = [0, 1, 2, 2, 4]`

`from_sequence[i] == ['T', 'T', 'C', 'G', 'T'] for each i`
`to_sequence[s1inds_gap_left[i]] == ['T', 'T', 'C', 'C', 'T'] for each i``

So they will always differ in the gap. Right?

Comment (Colelyman) [reply tests/unit_tests/test_CRISPRessoShared.py]: Yes, they will always differ at the gap. But they should point either to the left or right of the gap. Does that make sense?

File: tests/unit_tests/test_CRISPRessoCORE.py (hunk 1 above)
Comment (mbowcut2) [reply tests/unit_tests/test_CRISPRessoCORE.py]: @Colelyman update here: 

First, `-qwc` accepts 0 as part of a range and works correctly as long as `--exclude_bp_from_left` is set to 0 (the default is 15). 

Second, `-qwc` doesn't accept `0` as an argument. But it can be used by `-w` e.g. `-w 0`. It seems that `-qwc` supersedes other quantification window options, and so `-w 0 -qwc 10-20` would give you a window of 10-20.

Comment (mbowcut2) [reply tests/unit_tests/test_CRISPRessoCORE.py]: @Colelyman okay, it works now.

**Comment by Colelyman** [comment]
@mbowcut2 This looks good to me! I don't see anything else that needs to be changed, let's wait to merge it in until the other PRs are merged in first.

Thanks!
